File: week3/c++/4CloneGraph.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};


class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return nullptr;
        Node* cur = new Node(node->val);
        visited[node] = cur;
        for (auto neighbor : node->neighbors) {
            if (visited.find(neighbor) != visited.end()) {
                cur->neighbors.push_back(visited[neighbor]);
            }
            else {
                cur->neighbors.push_back(cloneGraph(neighbor));
            }
        }
        return cur;
    }
private:
    unordered_map<Node*, Node*> visited;
};

File: week3/Java/4CloneGraph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    private HashMap<Node, Node> map = new HashMap<>();
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        Queue<Node> q = new LinkedList<>();
        Node clone = new Node(node.val);
        map.put(node, clone);
        q.offer(node);
        while (!q.isEmpty()) {
            int qLen = q.size();
            for (int i = 0; i < qLen; i++) {
                Node cur = q.poll();
                for (Node next : cur.neighbors) {
                    if (!map.containsKey(next)) {
                        map.put(next, new Node(next.val));
                        q.add(next);
                    }
                    map.get(cur).neighbors.add(map.get(next));
                    
                }
            }
        }
        return clone;
    }
}